engine/src/main/scala/cromwell/engine/io/IoActor.scala

```diff
@@ -1,149 +1,212 @@
 package cromwell.engine.io
 
-import java.time.OffsetDateTime
-
 import akka.NotUsed
 import akka.actor.{Actor, ActorLogging, ActorRef, Props, Timers}
 import akka.dispatch.ControlMessage
 import akka.stream._
 import akka.stream.scaladsl.{Flow, GraphDSL, Merge, Partition, Sink, Source, SourceQueueWithComplete}
-import com.typesafe.config.ConfigFactory
+import com.typesafe.config.Config
 import cromwell.core.Dispatcher.IoDispatcher
 import cromwell.core.actor.StreamActorHelper
 import cromwell.core.actor.StreamIntegration.StreamContext
+import cromwell.core.io.Throttle._
 import cromwell.core.io.{IoAck, IoCommand, Throttle}
 import cromwell.core.{Dispatcher, LoadConfig}
 import cromwell.engine.instrumentation.IoInstrumentation
 import cromwell.engine.io.IoActor._
+import cromwell.engine.io.gcs.GcsBatchFlow.GcsBatchFlowConfig
 import cromwell.engine.io.gcs.{GcsBatchCommandContext, ParallelGcsBatchFlow}
 import cromwell.engine.io.nio.NioFlow
+import cromwell.engine.io.nio.NioFlow.NioFlowConfig
 import cromwell.filesystems.gcs.batch.GcsBatchIoCommand
 import cromwell.services.loadcontroller.LoadControllerService.{HighLoad, LoadMetric, NormalLoad}
+import net.ceedubs.ficus.readers.ValueReader
 
+import java.time.OffsetDateTime
+import java.time.temporal.ChronoUnit
 import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+
 
 /**
   * Actor that performs IO operations asynchronously using akka streams
   * 
-  * @param queueSize size of the queue
-  * @param throttle optional throttler to control the throughput of requests.
-  *                 Applied to ALL incoming requests
+  * @param ioConfig IoActor configuration class
   * @param materializer actor materializer to run the stream
   * @param serviceRegistryActor actorRef for the serviceRegistryActor
   */
-final class IoActor(queueSize: Int,
-                    nioParallelism: Int,
-                    gcsParallelism: Int,
-                    throttle: Option[Throttle],
+final class IoActor(ioConfig: IoConfig,
                     override val serviceRegistryActor: ActorRef,
                     applicationName: String)(implicit val materializer: ActorMaterializer)
   extends Actor with ActorLogging with StreamActorHelper[IoCommandContext[_]] with IoInstrumentation with Timers {
   implicit val ec: ExecutionContext = context.dispatcher
 
+  //noinspection ActorMutableStateInspection
```

Comment:
Suppressing this inspection makes me nervous, however I don't have the expertise with akka/akka-streams to weigh in further. Could you add a comment explaining your current understanding of why this mutable state is okay and/or why this inspection is falsely flagging this?

Reply:
I can add a comment. This is a fairly common pattern in Cromwell (the inspection is disabled in 5 other spots in the baseline code). IntelliJ seems to flag any mutable state in actors with this warning, but AFAIK it should be perfectly safe as long as the mutation only happens in `receive`.

Reply:
@breilly2 Rightly or wrongly - one of the use cases pitched for actors (at least in akka) is to use it as a way to isolate & protect mutable state. As @mcovarr notes - having `var` inside of actors is a common pattern. The flip side is this is (one reason) why the internals of an actor should be dead simple